test/e2e/tls_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// #nosec G101: false positive triggered by variable name which includes "token"
+const tokenFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+const cipherSuite = tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 // a TLS1.2 Cipher Suite
+
+var cipherSuites = []uint16{cipherSuite}
+
+// TestTLSCipherSuitesClient tests Cipher Suite and TLSVersion config on Antrea Apiserver.
+func TestTLSCipherSuitesAntrea(t *testing.T) {
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	data.configureTLS(t, cipherSuites, "VersionTLS12")
+	tls13CipherSuites := make(map[uint16]bool)
+	for _, cs := range tls.CipherSuites() {
+		if len(cs.SupportedVersions) == 1 && cs.SupportedVersions[0] == tls.VersionTLS13 {
+			tls13CipherSuites[cs.ID] = true
+		}
+	}
+
+	svc, err := data.clientset.CoreV1().Services(antreaNamespace).Get(context.TODO(), "antrea", metav1.GetOptions{})
+	assert.NoError(t, err, "failed to get Antrea Service")
+	if len(svc.Spec.Ports) == 0 {
+		t.Fatal("Antrea Service has no ports")
+	}
+	url := fmt.Sprintf("https://%s:%d", svc.Spec.ClusterIP, svc.Spec.Ports[0].Port)
+
+	// 1. TLSMaxVersion unset, then a TLS1.3 Cipher Suite should be used.
+	respCipherSuite := request(t, url, 0)
```

Comment:
you can't assume that the ClusterIP is accessible from outside the cluster, it's why the Kind tests are failing

you can use `openssl` like you did for the Agent apiserver below. Or maybe you can get the Pod IP for the antrea-agent / antrea-controller Pod, which would be the Node IP, since these are hostNetwork Pods. It's safe to assume that the Node IP can be reached from the test program, since we support running commands on Nodes over SSH. Either way, maybe we should unify the "request" method for the Controller apiserver and the Agent apiserver.

Reply:
You're right. Code refactored.